_posts/2016-02-07-coupling-in-distributed-systems.md

```diff
@@ -0,0 +1,88 @@
+
+I would prefer spending my time creating tests under the [motivations](https://twitter.com/sarahmei/status/685907333889810432) of providing living documentation, easing future refactors or helping me to drive the design, more than helping catching bugs that the type system should take care of. Writing a test that checks the behaviour of a method when receives null it's a waste of time if we can make impossible to write a line of code that passes a null.
+
+Compile world is not perfect though, as it's definitively slower on development and constrains developers (someone could say that less freedom might be a nice to have in this context)
+
+## Runtime approaches
+
+Now that I've been honest with you about my compile bias I can explain different approaches and trade-offs for the schema validation problem.
+
+### Implicit schemas
+
+First runtime approach is the loosest one: using implicit schemas and trusting in the good will of producers. As nobody is checking the validity of messages before of being published into Kafka that means that consumers could blow up.
+
+First corrective measure is assuring that only the processing of the poisoned message will blow and not the whole consumer. An example of that would be providing a resume supervision strategy on [Akka Streams](http://doc.akka.io/docs/akka-stream-and-http-experimental/2.0.2/scala.html) when the message don't hold expected implicit schema.
+
+Second corrective measure would be not simply swallowing those crashes but being able to communicate them to proper actors (being humans or software). Good practice is to provide dead letter queues for poisoned messages just in case we want to manipulate and retry the processing of those messages at that level.
+
+Before of getting into explicit schemas I would say that those measure are not usually enough but they are a good safety net, as shit happens, and we need to be prepared.
+
+### Explicit schemas
+
+If we want to avoid poisoned messages to get into our topics we could provide a middle-man service to intercept and validate explicit schemas. That's an integration service that could be a single point of failure but, at the same time, it could valuable to have a centralised repo of schemas when we have a lot of consumers and the complexity of the system would be hard to grasp in de-centralised fashion.
+
+[Schema registry](http://docs.confluent.io/1.0/schema-registry/docs/index.html) is an example of that for Kafka, and its documentation is full of insights about how to implement that in a distributed, highly available and scalable way.
+
+## Compile approaches
+
+Last approach would be creating software that makes impossible to create messages that not hold the expected schema. Assuming [Scala](http://www.scala-lang.org/), we could create a jar that contains [case classes](http://docs.scala-lang.org/tutorials/tour/case-classes.html) that are object materialisations of a schema.
```

Comment:
`that makes impossible to create messages that not hold the expected schema` -> `that makes it impossible to create message that do not hold the expected schema`